include/swift/Syntax/SyntaxArena.h

```diff
@@ -41,8 +51,42 @@ class SyntaxArena : public llvm::ThreadSafeRefCountedBase<SyntaxArena> {
 public:
   SyntaxArena() {}
 
+  ~SyntaxArena() {
+    // Release all child arenas that were manually retained in addChildArena.
+    for (SyntaxArena *ChildArena : ChildArenas) {
+      ChildArena->Release();
+    }
+  }
+
   static RC<SyntaxArena> make() { return RC<SyntaxArena>(new SyntaxArena()); }
 
+  /// Add an arena that is kept alive while this arena lives. See documentation
+  /// of \c ChildArenas for more detail.
+  void addChildArena(SyntaxArena *Arena) {
+    if (Arena == this) {
+      return;
+    }
+    auto DidInsert = ChildArenas.insert(Arena);
+    if (DidInsert.second) {
+      Arena->Retain();
+      assert(!containsReferenceCycle());
```

Comment:
```suggestion
      assert(!Arena->containsReferenceCycle({this}));
```

Reply:
Nice and small performance improvement. I like it 👍🏽